topology2: Mod cfg ibs obs cpc addition

jsarha opened the pull request:
I am feeling quite clueless here, so even if this does not work, I push this out in hope someone can spot the problem.

For some reason the widget tuple array only contains the first instance of mod_cfg objects in the topology2 conf file. I am trying to read the array out with this code: https://github.com/jsarha/linux-sof/tree/peter-cpc-test

lgirdwood commented on tools/topology/topology2/cavs-mixin-mixout-hda.conf:
```diff
@@ -36,11 +36,34 @@ Object.Pipeline {
+					obs 3145
+					cpc 3178
+				}
+				num_mod_cfgs 3
```
Do we need this, the CC should count this.

jsarha replied:
Ok. I was actually wondering why some widgets have explicitly set num_audio_formats and some doesn't.

ranj063 replied:
some widgets support only 1 format, so they're set in teh class definition. Others support variable number of format based on tplg, so they're set in the top-level file

jsarha commented:
@ranj063 can help me out here?

ranj063 commented:
> I am feeling quite clueless here, so even if this does not work, I push this out in hope someone can spot the problem.
> 
> For some reason the widget tuple array only contains the first instance of mod_cfg objects in the topology2 conf file. I am trying to read the array out with this code: https://github.com/jsarha/linux-sof/tree/peter-cpc-test

@jsarha the tplg is fine. but what you are missing is the parsing bits in the kernel. Take a look at how we parset multiple audio formats here: https://github.com/thesofproject/linux/blob/bef9852fe3228ce83717dada3ebcb7ad07c6099b/sound/soc/sof/topology.c#L1238

You need to allocate memory for "n" number of token sets first. And then parse the number of sets you're expecting like https://github.com/thesofproject/linux/blob/bef9852fe3228ce83717dada3ebcb7ad07c6099b/sound/soc/sof/ipc4-topology.c#L221 and then finally parse the tokens like https://github.com/thesofproject/linux/blob/bef9852fe3228ce83717dada3ebcb7ad07c6099b/sound/soc/sof/ipc4-topology.c#L257. Hope this helps

lgirdwood commented:
@jsarha @ujfalusi @ranj063 we need to have a per HW target per module file for each module i.e.
```
cavs/tgl/copier.conf
ace/mtl/copier.conf
```
and so on that we would include in teh high level generic pipeline files.

jsarha commented:
>

sys-pt1s commented:
Can one of the admins verify this patch?

kv2019i commented:
@jsarha Status? Please mark as draft if this is not intended for merging...

jsarha commented:
> @jsarha Status? Please mark as draft if this is not intended for merging...

This PR is good as such, but it does not contain the structures requested here: https://github.com/thesofproject/sof/pull/7348#issuecomment-1497387922

They could be implemented as separate PR. But maybe I'll mark this as draft as you requested and try to put all the pieces together. Then again I have no idea of any mod_cfg settings for any platform, so in the end there will not be any settings in the later PR either, just places for different platforms to put them.

jsarha commented:
Actually, I am not so sure about the second commit in this series. Maybe its better to include the mod_cfg.conf already in the component files (e.g. in include/component/*.conf), and not in the pipeline definitions. I already have an alternative commit: https://github.com/thesofproject/sof/commit/ab8538cb37559fb477d1e5c8878d634b71cbf68f  
@ranj063 , this is better, isn't it?

ranj063 commented:
> Actually, I am not so sure about the second commit in this series. Maybe its better to include the mod_cfg.conf already in the component files (e.g. in include/component/*.conf), and not in the pipeline definitions. I already have an alternative commit: [ab8538c](https://github.com/thesofproject/sof/commit/ab8538cb37559fb477d1e5c8878d634b71cbf68f) @ranj063 , this is better, isn't it?

@jsarha yes adding it to the widget class definitions where they will be instantaited and used is probably better

jsarha commented:
New version pushed. This version should have what was requested in https://github.com/thesofproject/sof/pull/7348#issuecomment-1497387922 . Please check that I have added the placeholder and included it in actual component widgets. I already removed pipeline from them, but for many of them I have no idea what those entities are.

jsarha commented:
> @jsarha conflicts?

Solved now.

ranj063 commented on tools/topology/topology2/include/components/asrc.conf:
```diff
@@ -29,6 +29,12 @@ Class.Widget."asrc" {
 	#include common component definition
 	<include/components/widget-common.conf>
 
+	#include platform specific module configs for the asrc module
```
@jsarha is there value in this commit just yet? Can we not add the files as and when needed?

kv2019i replied:
@lgirdwood Can you ack as this was an ask from you? I.e. to have the placeholders or not.

btian1 replied:
1. I agree add on demand.
2. if want to add in this patch, I would suggest there is a folder to cover all component instead of directly under platform/intel/mod_cfg/

kv2019i reviewed (approved):
Not sure what is more likely, need to override values per platform, or per specific product, but in any case, this PR provides the necessary building blocks for both needs, so seems good.

jsarha commented:
Pushed forwards, and changes requested in mail, so turning to draft again.

Girdwood, Liam R:
> Jyri, a topology override must copy the manifest CPC schema exactly.
> Even the unused fields must be copied and populated with 0 today.

lgirdwood commented:
@jsarha now that the manifest and kernel CPC parts are working we should add the overide option for v2.7.
@ujfalusi @plbossart fyi.

plbossart commented:
> @jsarha now that the manifest and kernel CPC parts are working we should add the overide option for v2.7. @ujfalusi @plbossart fyi.

We need to agree on the definition of 'working'. I still routinely see messages telling me that CPC is not set ?

lgirdwood commented:
> > @jsarha now that the manifest and kernel CPC parts are working we should add the overide option for v2.7. @ujfalusi @plbossart fyi.
> 
> We need to agree on the definition of 'working'. I still routinely see messages telling me that CPC is not set ?

There are still a lot of holes in the manifest CPC data for each module and (format, rate, channels) configuration. @aiChaoSONG will be working on populating this prior to v2.7. This should fix the kernel warning messages for all tested configs.

The toplogy part will use the same CPC schema and kernel logic/IPC, but will allow 
1) Adding a missing CPC data element if one is missing in manifest for a new configuration.
2) Overide an existing manifest CPC data element if needed.

plbossart commented:
I'd like to reach the point where all CPC data is valid in the manifest and used for something meaningful, before we plan work on overriding stuff.

lgirdwood commented:
> I'd like to reach the point where all CPC data is valid in the manifest and used for something meaningful, before we plan work on overriding stuff.

Need to do both in parallel, to be ready for mistakes in manifest CPC.

plbossart commented:
> > I'd like to reach the point where all CPC data is valid in the manifest and used for something meaningful, before we plan work on overriding stuff.
> 
> Need to do both in parallel, to be ready for mistakes in manifest CPC.

we can't afford to hedge our bets for every single feature. The firmware team is responsible for providing correct CPC values, if they don't it's a bug and they fix it. 
 We have lots of other things to do really with SoundWire-related enablement.

lgirdwood commented:
> > > I'd like to reach the point where all CPC data is valid in the manifest and used for something meaningful, before we plan work on overriding stuff.
> > 
> > 
> > Need to do both in parallel, to be ready for mistakes in manifest CPC.
> 
> we can't afford to hedge our bets for every single feature. The firmware team is responsible for providing correct CPC values, if they don't it's a bug and they fix it. We have lots of other things to do really with SoundWire-related enablement.

CPC architecture is based on a single module unit test with no upstream/downstream module I$/D$ impact taken into consideration for module CPC performance delta when different pipeline topologies are used. This is not a big item and will be painful to modify/fix in manifest.

kv2019i commented:
In draft status, pushing to v2.9.

btian1 commented on tools/topology/topology2/include/common/mod_cfg.conf:
```diff
@@ -0,0 +1,45 @@
+# Object.Base.mod_cfg."0" {
+#	ibs	384
+#	obs	384
+#	cpc	4600
```
can we add comments here:
for ibs/obs, it is initialized as: 48 * 2 * 4 = 384, it is 48k LL with 2 channel 32 bit format.
for cpc, I am not sure how CPC figured out as instantiated value?

btian1 commented on tools/topology/topology2/include/common/mod_cfg.conf:
```diff
@@ -0,0 +1,45 @@
+    DefineAttribute."instance" {
+    }
+
+    # input buffer size
```
let's give comments like: 
ibs = (sample rate / (1000000/period)) * channels * valid_sample_width

btian1 commented on tools/topology/topology2/include/components/widget-common-inc.conf:
```diff
@@ -0,0 +1,5 @@
+# This file is for include files needed for widget-common.conf
+#
+<include/common/mod_cfg.conf>
+
```
why not directly include mod_cfg.conf?

kv2019i commented:
Stable-v2.9 branched, this didn't make the cut, bumping forward. @jsarha is this still valid (given bumps over multiple releases)?

kv2019i commented:
No activity in 2.7...2.9, clearing the milestone.
